internal/model/modelindexer/indexer.go

```diff
@@ -234,54 +235,21 @@ func (i *Indexer) processEvent(ctx context.Context, event *model.APMEvent) error
 	r.indexBuilder.WriteString(event.DataStream.Dataset)
 	r.indexBuilder.WriteByte('-')
 	r.indexBuilder.WriteString(event.DataStream.Namespace)
-	index := r.indexBuilder.String()
-
-	i.activeMu.Lock()
-	defer i.activeMu.Unlock()
-	for i.active != nil && i.active.Len() >= i.config.FlushBytes {
-		// The active bulk indexer is full: wait for it to be
-		// switched out by the background flushActive goroutine.
-		i.activeCond.Wait()
-	}
-	if i.active == nil {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case i.active = <-i.available:
-			atomic.AddInt64(&i.availableBulkRequests, -1)
-		}
-		if i.timer == nil {
-			i.timer = time.NewTimer(i.config.FlushInterval)
-		} else {
-			i.timer.Reset(i.config.FlushInterval)
-		}
-		i.g.Go(func() error {
-			// The timer may be stopped by i.Close or when
-			// i.config.FlushBytes is exceeded, in which case
-			// i.timerStopped will be signalled.
-			select {
-			case <-i.timerStopped:
-			case <-i.timer.C:
-			}
-			return i.flushActive(context.Background())
-		})
-	}
 
-	if err := i.active.Add(elasticsearch.BulkIndexerItem{
-		Index:  index,
+	select {
+	// Send the BulkIndexerItem to the internal channel, allowing individual
+	// events to be processed by an active bulk indexer in a dedicate goroutine,
+	// which in turn speeds up event processing.
+	case i.bulkItems <- elasticsearch.BulkIndexerItem{
+		Index:  r.indexBuilder.String(),
 		Action: "create",
 		Body:   r,
-	}); err != nil {
-		return err
+	}:
```

Comment:
That makes sense, however, for better or worse we'd be performing that processing by the consumer which was a more limited number of concurrency (1 at the moment of this PR), rather than the producer (up to 200 concurrent json encodes). This could slow down the pipeline, but if we implement multiple active indexers, then I think that makes a lot of sense since we'd have more than a single consumer

Reply:
I think we should still try and measure it later. It may be that removing the intermediate buffer & copying offsets the loss of concurrency. Anyway, not for now.